### ISSUE TITLE
tasks

### ISSUE THREAD

**Issue by 95155833** [95155833/ai-projects#1]
import cv2
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import load_model
from tensorflow.keras.preprocessing.image import img_to_array
import os

face_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')
recognition_model = load_model('path/to/face_recognition_model.h5')

def preprocess_image(img, target_size=(160, 160)):
    img = cv2.resize(img, target_size)
    img = img_to_array(img)
    img = np.expand_dims(img, axis=0)
    img = img / 255.0
    return img

def recognize_face(model, face):
    face = preprocess_image(face)
    embedding = model.predict(face)[0]
    return embedding

def load_known_faces(directory):
    known_faces = {}
    for filename in os.listdir(directory):
        if filename.endswith(".jpg") or filename.endswith(".png"):
            img_path = os.path.join(directory, filename)
            img = cv2.imread(img_path)
            face = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
            face_id = filename.split('.')[0]
            known_faces[face_id] = recognize_face(recognition_model, face)
    return known_faces

def find_closest_face(known_faces, embedding):
    min_dist = float('inf')
    identity = None
    for name, known_embedding in known_faces.items():
        dist = np.linalg.norm(embedding - known_embedding)
        if dist < min_dist:
            min_dist = dist
            identity = name
    return identity, min_dist

known_faces = load_known_faces('path/to/known_faces')

def process_frame(frame):
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    faces = face_cascade.detectMultiScale(gray, 1.3, 5)
    for (x, y, w, h) in faces:
        roi_gray = gray[y:y+h, x:x+w]
        embedding = recognize_face(recognition_model, roi_gray)
        identity, distance = find_closest_face(known_faces, embedding)
        label = identity if distance < 0.6 else "Unknown"
        cv2.rectangle(frame, (x, y), (x+w, y+h), (255, 0, 0), 2)
        cv2.putText(frame, label, (x, y-10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (255, 0, 0), 2)
    return frame

cap = cv2.VideoCapture(0)
while True:
    ret, frame = cap.read()
    if not ret:
        break
    frame = process_frame(frame)
    cv2.imshow('Face Detection and Recognition', frame)
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break
cap.release()
cv2.destroyAllWindows()
